cheatsheets/gleam-for-go-users.md

```diff
@@ -0,0 +1,1447 @@
+static class.
+
+In comparison Gleam modules can also contain custom types.
+
+A gleam module name corresponds to its file name and path.
+
+Since there is no special syntax to create a module, there can be only one
+module in a file and since there is no way name the module the filename
+always matches the module name which keeps things simple and transparent.
+
+In `/src/something/somethingElse.gleam`:
+
+```gleam
+// Creation of module function identity
+// in module somethingElse
+pub fn identity(x) {
+  x
+}
+```
+
+Importing the `somethingElse` module and calling a module function:
+
+```gleam
+// In src/main.gleam
+import something/somethingElse // if something was in a directory called `lib` the import would be `lib/something/somethingElse`.
+
+pub fn main() {
+  somethingElse.identity(1) // 1
+}
+```
+
+### Imports
+
+#### Go
+
+Go imports are fairly simple.  Using `import` keyword with either a string literal
+that contains the path to a package, or several such strings each on separate lines
+inside of parentheses.
+
+Besides the standard library, all packages have to be fully qualified, where the
+qualifier is the name of the containing module, which by convention is the url
+at which the module is hosted.
+
+```go
+import "fmt" // Single import.	Stdlib packages are referenced bare, no path required
+import ( // Multi-package import
+	"strings" // Available for use as `strings`
+
+	"github.com/example/example-go-project/somepkg" // Will be available as `somepkg`
+)
+```
+
+There is no way in Go to import a symbol itself, all imports are qualified by their
+package.
+
+```go
+fmt.Println()
+strings.Trim()
+```
+
+#### Gleam
+
+Imports are relative to the app `src` folder.
+
+Modules in the same directory will need to reference the entire path from `src`
+for the target module, even if the target module is in the same folder.
+
+Inside module `src/nasa/moon_base.gleam`:
+
+```gleam
+// imports module src/nasa/rocket_ship.gleam
+import nasa/rocket_ship
+
+pub fn explore_space() {
+  rocket_ship.launch()
+}
+```
+
+### Named imports
+
+#### Go
+
+Go allows you to alias a package import
+
+```go
+import (
+	other "github.com/example/example-go-project/somepkg"
+)
+
+func main() {
+	other.SomeFunc()
+}
+```
+
+#### Gleam
+
+Gleam has as similar feature:
+
+```gleam
+import unix/cat
+import animal/cat as kitty
+// cat and kitty are available
+```
+
+This may be useful to differentiate between multiple modules that would have the same default name when imported.
+
+### Unqualified imports
+
+#### Go
+
+Go does not support unqualified imports
+
+#### Gleam
+
+```gleam
+import animal/cat.{
+  Cat,
+  stroke
+}
+
+pub fn main() {
+  let kitty = Cat(name: "Nubi")
+  stroke(kitty)
+}
+```
+
+Importing common types such as `gleam/order.{Lt, Eq, Gt}` or
+`gleam/option.{Some,None}` can be very helpful.
+
+## Architecture
+
+To iterate a few foundational differences (Go VS Gleam):
+
+1. Programming model: imperative VS functional immutable
```

Comment:
A common Go-ism is something like:

- Declare struct
- Declare constructor
- Declare methods on struct

It would be handy to see how you'd achieve this in Gleam. 

I know that for Gleam some people recommend a "context" struct that you pass in as the first argument which contains what would normally be Go struct fields: Database connections, loggers, HTTP clients etc.